client/mysqltest.cc

```diff
@@ -3837,44 +3837,337 @@ void do_remove_files_wildcard(struct st_command *command)
+
+  Copy will fail if no files match the pattern. Copy will also
+  fail if source directory or destination directory or both do not
+  exist.
+
+  ds_source      - pointer to dynamic string containing source
+                   directory information
+  ds_destination - pointer to dynamic string containing destination
+                   directory information
+  ds_wild        - pointer to dynamic string containing wildcard pattern
+  recursive      - indicating whether to copy directory recursively
+*/
+static int wildcard_copy(DYNAMIC_STRING *ds_source,
+                         DYNAMIC_STRING *ds_destination,
+                         DYNAMIC_STRING *ds_wild,
+                         bool recursive)
+{
+  int error= 0;
+
+  /* Note that my_dir sorts the list if not given any flags */
+  MY_DIR *src_dir_info= my_dir(ds_source->str,
+                               MYF(MY_DONT_SORT | MY_WANT_STAT));
+  MY_DIR *dst_dir_info= my_dir(ds_destination->str,
+                               MYF(MY_DONT_SORT | MY_WANT_STAT));
+
+  /* Either one of directories does not exist or access denied */
+  if (!src_dir_info || !dst_dir_info)
+  {
+    error= 1;
+    goto end;
+  }
+
+  char dir_separator[2];
+  dir_separator[0]= FN_LIBCHAR;
+  dir_separator[1]= 0;
+  dynstr_append_mem(ds_source, dir_separator, strlen(dir_separator));
+  dynstr_append_mem(ds_destination, dir_separator, strlen(dir_separator));
+
+  /* Storing the length of the path to the file, so it can be reused */
+  size_t source_file_length;
+  size_t dest_file_length;
+  dest_file_length= ds_destination->length;
+  source_file_length= ds_source->length;
+  uint match_count;
+  match_count= 0;
+
+  for (uint i= 0; i < src_dir_info->number_of_files; i++)
+  {
+    ds_source->length= source_file_length;
+    ds_destination->length= dest_file_length;
+    FILEINFO *file= src_dir_info->dir_entry + i;
+
+    /* Skip directories if recursive copy is not requested */
+    if (MY_S_ISDIR(file->mystat->st_mode) && !recursive)
+      continue;
+
+    /* Copy only those files which the pattern matches */
+    if (ds_wild->length &&
+        wild_compare(file->name, ds_wild->str, false))
+      continue;
+
+    match_count++;
+    dynstr_append_mem(ds_source, file->name, strlen(file->name));
+    dynstr_append_mem(ds_destination, file->name, strlen(file->name));
+
+    if (MY_S_ISDIR(file->mystat->st_mode))
+    {
+      DBUG_PRINT("info", ("Recursive copy files of %s to %s",
+                          ds_source->str, ds_destination->str));
+      DBUG_PRINT("info", ("listing directory: %s", ds_source->str));
+      error= recursive_copy(ds_source, ds_destination);
+    }
+    else
+    {
+      DBUG_PRINT("info", ("Copying file: %s to %s",
+                          ds_source->str, ds_destination->str));
+      /* MY_HOLD_ORIGINAL_MODES prevents attempts to chown the file */
+      error= (my_copy(ds_source->str, ds_destination->str,
+              MYF(MY_HOLD_ORIGINAL_MODES)) != 0);
```

Comment:
I don't quite like that wildcard copy will automatically overwrite and non-wildcard copy won't.
Perhaps you should also introduce -f flag? For simplicity, let's only support bundling, that is, one cannot write "-r -f" but only "-rf".
Then you'll have, instead of
```c
   recursive= !strcmp("-r", ds_recursive.str);
```
something like
```c
  force= MY_DONT_OVERWRITE_FILE;
  if (ds_recursive.str[0] == '-')
    for (i=1; i < ds_recursive.length; i++)
      if (ds_recursive.str[i] == 'r')
        recursive= 1;
      else if (ds_recursive.str[i] == 'f')
        force= 0;
      else syntax_error;
  else syntax_error;
```
and you can simply OR the `force` flag with other flags in `my_copy()`.